Work well with older `lock_api` versions

Kijewski opened the pull request:
The crate `lock_api` is a transitive dependency, pulled in trough the use of `parking_lot`. `parking_lot` depends on `lock_api` version `^0.4.6`, but only since `lock_api` 0.4.7 the `RwLock::new()` method is available on rust stable, coinciding with the release of rust [1.61][const_fn_trait_bound].

A problem occurs if you try to use `once_map` with non-up-to-date dependencies, e.g. by running

```sh
RUSTC_BOOTSTRAP=1 cargo generate-lockfile -Zminimal-versions
```

~~This PR replaces the use of `RwLock::new()` with `RwLock::const_new()`, which was already available in `lock_api` 0.4.6.~~

This PR makes `Shard::new()` non-`const`, so it works with older `lock_api` versions, too.

[const_fn_trait_bound]: <https://caniuse.rs/features/const_fn_trait_bound>

Kijewski commented:
And thank you for your fine library! The API of e.g. [`OnceMap::get_or_try_insert_ref()`](https://docs.rs/once_map/0.4.18/once_map/sync/struct.OnceMap.html#method.get_or_try_insert_ref) is really well designed, and really easy to use.

a1phyr commented:
Thanks for the PR!

I would prefer to require `lock_api >= 0.4.7` (which is compatible with our MSRV), or even simpler, remove the `const` attribute which is unused anyway.

Kijewski committed:
Work well with older `lock_api` versions

The crate `lock_api` is a transitive dependency, pulled in trough the
use of `parking_lot`. `parking_lot` depends on `lock_api` version
`^0.4.6`, but only since `lock_api` 0.4.7 the `RwLock::new()` method is
available on rust stable, coinciding with the release of rust
[1.61][const_fn_trait_bound].

A problem occurs if you try to use `once_map` with non-up-to-date
dependencies, e.g. by running

```sh
RUSTC_BOOTSTRAP=1 cargo generate-lockfile -Zminimal-versions
```

This PR makes `Shard::new()` non-`const`, so it works with older
`lock_api` versions, too.

Kijewski commented:
I chose the explicit dependency route, so it's not a breaking change. I think `const` constructor can come in handy in some circumstances.

a1phyr commented:
This function is not public, so it would not break anything

Kijewski commented:
Oops, yes, you're right! I assumed it was used to implement some other function as `const`.

a1phyr commented:
Thanks a lot !
